modules/mod-deep-learning/DeepLearningEffect.cpp

```diff
@@ -0,0 +1,182 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepLearningEffect.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+/**
+
+*/
+/*******************************************************************/
+
+#include "DeepLearningEffect.h"
+
+#include <wx/stattext.h>
+
+#include "FileNames.h"
+#include "Shuttle.h"
+#include "ShuttleGui.h"
+#include "WaveTrack.h"
+
+#include "effects/LoadEffects.h"
+
+#include <torch/script.h>
+
+// DeepLearningEffectBase implementation
+
+std::string DeepLearningEffect::GetDeepEffectID()
+{ return "waveform-to-waveform";}
+
+const ComponentInterfaceSymbol DeepLearningEffect::Symbol
+{ XO("Deep Learning Effect") };
+
+// register source separation
+namespace{ BuiltinEffectsModule::Registration< DeepLearningEffect > reg; }
+
+DeepLearningEffect::DeepLearningEffect()
+{
+   SetLinearEffectFlag(false);
+}
+
+DeepLearningEffect::~DeepLearningEffect()
+{
+}
+
+// ComponentInterface implementation
+
+ComponentInterfaceSymbol DeepLearningEffect::GetSymbol()
+{
+   return Symbol;
+}
+
+TranslatableString DeepLearningEffect::GetDescription()
+{
+   return XO("The goal of audio source separation is to isolate "
+             "the sound sources in a given mixture of sounds.");
+}
+
+ManualPageID DeepLearningEffect::ManualPage()
+{
+   return L"Source_Separation"; 
+}
+
+// EffectDefinitionInterface implementation
+
+EffectType DeepLearningEffect::GetType()
+{
+   return EffectTypeProcess;
+}
+
+// Effect implementation
+
+// ProcessOne() takes a track, transforms it to bunch of buffer-blocks,
+// performs a forward pass through the deep model, and writes 
+// the output to new tracks. 
+bool DeepLearningEffect::ProcessOne(WaveTrack *leader,
+                           double tStart, double tEnd)
+{
+   if (leader == NULL)
+      return false;
+
+   // keep track of the sample format and rate,
+   // we want to convert all output tracks to this
+   sampleFormat origFmt = leader->GetSampleFormat();
+   int origRate = leader->GetRate();
+
+   // initialize source tracks, one for each source that we will separate
+   std::vector<WaveTrack::Holder> sourceTracks;
+   std::vector<std::string> sourceLabels = mModel->GetCard()->labels();
+   sourceTracks = CreateSourceTracks(leader, sourceLabels);
```

Comment:
As many tracks are made, as there are labels.  But not, number of lables times number of channels in the leader.

Is this right?  Should there be more commentary about how stereo tracks are handled?  Does a stereo-capable model "label" left and right outputs separately?  If so, then how do you inform the model that the input is not stereo?  (By a tensor dimension?)

Are you making any hidden assumptions that the input is at most stereo?  That is correct for now, but who knows how Audacity might generalize in the future.  If the assumption is important, put in a comment like other comments elsewhere in the source tree:

`// TODO: more-than-two-channels`

Reply:
yeah, the input to the deep model can currently have any number of channels (more-than-two-channels is supported), but each output track has to be mono. I don't think it would be too hard to support multichannel output tracks though.